sycl/unittests/event/EventDestruction.cpp

```diff
@@ -0,0 +1,208 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include <helpers/CommonRedefinitions.hpp>
+#include <helpers/PiMock.hpp>
+#include <helpers/TestKernel.hpp>
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+
+using namespace sycl;
+
+static int ReleaseCounter = 0;
+static pi_result redefinedEventRelease(pi_event event) {
+  ++ReleaseCounter;
+  return PI_SUCCESS;
+}
+
+pi_result redefinedMemBufferCreate(pi_context, pi_mem_flags, size_t size,
+                                   void *, pi_mem *,
+                                   const pi_mem_properties *) {
+  return PI_SUCCESS;
+}
+
+class EventDestructionTest : public ::testing::Test {
+public:
+  EventDestructionTest() : Plt{default_selector()} {}
+
+protected:
+  void SetUp() override {
+    if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+      std::cout << "This test is only supported on OpenCL backend\n";
+      std::cout << "Current platform is "
+                << Plt.get_info<sycl::info::platform::name>();
+      return;
+    }
+
+    Mock = std::make_unique<unittest::PiMock>(Plt);
+
+    setupDefaultMockAPIs(*Mock);
+    Mock->redefine<detail::PiApiKind::piEventRelease>(redefinedEventRelease);
+    Mock->redefine<sycl::detail::PiApiKind::piMemBufferCreate>(
+        redefinedMemBufferCreate);
+  }
+
+protected:
+  std::unique_ptr<unittest::PiMock> Mock;
+  platform Plt;
+};
+
+// Test that events are destructed in correct time
+TEST_F(EventDestructionTest, EventDestruction) {
+  if (Plt.is_host() || Plt.get_backend() != backend::opencl) {
+    return;
+  }
+  sycl::context Context{Plt};
+  sycl::queue Queue{Context, sycl::default_selector{}};
+
+  {
+    ReleaseCounter = 0;
+    sycl::event E1{};
+
+    {
+      sycl::event E0 = Queue.submit([&](cl::sycl::handler &cgh) {
+        cgh.single_task<TestKernel>([]() {});
+      });
+      E1 = Queue.submit([&](cl::sycl::handler &cgh) {
+        cgh.depends_on(E0);
+        cgh.single_task<TestKernel>([]() {});
+      });
+      E1.wait();
+    }
+    // When a command is cleared we clear now only dependencies of the
+    // dependencies of the associated event. So, when the command
+    // associated with E0 event is destroyed, this event is still in
+    // E1 dependencies, which will not be cleared.
+    // Therefore no event release should be called until here.
+    EXPECT_EQ(ReleaseCounter, 0);
+
+    sycl::event E2 = Queue.submit([&](cl::sycl::handler &cgh) {
+      cgh.depends_on(E1);
+      cgh.single_task<TestKernel>([]() {});
+    });
+    E2.wait();
+    // Dependencies of E1 should be cleared here. It depends on E0.
+    EXPECT_EQ(ReleaseCounter, 1);
+
+    sycl::event E3 = Queue.submit([&](cl::sycl::handler &cgh) {
+      cgh.depends_on({E1, E2});
+      cgh.single_task<TestKernel>([]() {});
+    });
+    E3.wait();
+    // Dependency of E1 has already cleared. E2 depends on E1 that
+    // can't be cleared yet.
+    EXPECT_EQ(ReleaseCounter, 1);
+  }
```

Comment:
Upon leaving this scope E1 is going to be released.
Should this be checked?

Reply:
I think it is not necessary because after this scope all of the events will be destroyed in usual order.

Reply:
What is 'usual' order?

Reply:
I mean all of the events just will be destroyed